src/Microsoft.ML.Data/Transforms/ValueMappingTransformer.cs

```diff
@@ -153,6 +153,23 @@ private static ValueGetter<VBuffer<ReadOnlyMemory<char>>> GetKeyValueGetter<TKey
                 };
         }
 
+        internal static void AddColumnWrapper<T>(ArrayDataViewBuilder builder, string columnName, PrimitiveType primitiveType, T[] values)
+        {
+            if (typeof(T) == typeof(string))
+                builder.AddColumn(columnName, values.Select(x=>x.ToString()).ToArray());
```

Comment:
>values.Select(x=>x.ToString() [](start = 46, length = 29)

 `T[] values`
`typeof(T) == typeof(string)`

why you need to call ToString() on string?

Reply:
Because I need to convert to a memory stream with AsMemory, which  needs to operate on a string type. .ToString ensures that whatever type T is, it will be done on a string.

---
In reply to: [246978062](https://github.com/dotnet/machinelearning/pull/2115#discussion_r246978062) [](ancestors = 246978062)

Reply:
but this condition will get hit only if typeof(T) == typeof(string); so you know for sure values is a string.. 

---
In reply to: [247232387](https://github.com/dotnet/machinelearning/pull/2115#discussion_r247232387) [](ancestors = 247232387,246978062)

Reply:
i think that the if clause, will not execute, if T is not a string, it will fall to else.. 

---
In reply to: [247605115](https://github.com/dotnet/machinelearning/pull/2115#discussion_r247605115) [](ancestors = 247605115,247304740,247232387,246978062)

Reply:
I don't quite follow if you are saying that falling to the else is a problem or if the compiler should deduce that T is of type string.

Overall I am wanting to handle the case where T is of type string. Everything else should fall to the else clause as we would handle it as-is. In the case when T is of type string, we still have to convert T to a string because I am not able to call .AsMemory on a T object.  Therefore I am ensuring it is a string type by calling toString (hopefully calling toString on a string object is a no-op -- and this will only happen on strings because of the if-check). The other option is that I can cast T to a string if that is more clear.

Does that address your concern? Or am I misunderstanding?


---
In reply to: [249960836](https://github.com/dotnet/machinelearning/pull/2115#discussion_r249960836) [](ancestors = 249960836,247605115,247304740,247232387,246978062)